crates/dyn-abi/src/event.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{DynSolType, DynSolValue, Error, Result};
+use alloc::vec::Vec;
+use alloy_primitives::{Log, B256};
+
+/// A dynamic ABI event.
+///
+/// This is a representation of a Solidity event, which can be used to decode
+/// logs.
+#[derive(Debug, Clone, PartialEq)]
+pub struct DynSolEvent {
+    pub(crate) topic_0: Option<B256>,
+    pub(crate) indexed: Vec<DynSolType>,
+    pub(crate) body: DynSolType,
+}
+
+impl DynSolEvent {
+    /// Creates a new event, without length-checking the indexed, or ensuring
+    /// the body is a tuple. This allows creation of invalid events.
+    pub fn new_unchecked(
+        topic_0: Option<B256>,
+        indexed: Vec<DynSolType>,
+        body: DynSolType,
+    ) -> Self {
+        Self {
+            topic_0,
+            indexed,
+            body,
+        }
+    }
+
+    /// Creates a new event.
+    ///
+    /// Checks that the indexed length is less than or equal to 4, and that the
+    /// body is a tuple.
+    pub fn new(topic_0: Option<B256>, indexed: Vec<DynSolType>, body: DynSolType) -> Option<Self> {
+        if indexed.len() > 4 || body.as_tuple().is_none() {
+            return None
+        }
+        Some(Self::new_unchecked(topic_0, indexed, body))
+    }
+
+    /// True if anonymous.
+    pub const fn is_anonymous(&self) -> bool {
+        self.topic_0.is_none()
+    }
+
+    /// Decode the event from the given log info.
+    pub fn decode_log_parts<I>(
+        &self,
+        topics: I,
+        data: &[u8],
+        validate: bool,
+    ) -> Result<DecodedEvent>
+    where
+        I: IntoIterator<Item = B256>,
+    {
+        let mut topics = topics.into_iter();
+        let num_topics = self.indexed.len() + !self.is_anonymous() as usize;
+        if validate {
+            match topics.size_hint() {
+                (n, Some(m)) if n == m && n != num_topics => {
+                    return Err(Error::TopicLengthMismatch {
+                        expected: num_topics,
+                        actual: n,
+                    })
+                }
+                _ => {}
+            }
+        }
```

Comment:
👍

Reply:
this is @DaniPopes code I just rearranged it